scripts/k8s-diff/README.md

```diff
@@ -0,0 +1,274 @@
+Instead, this tool is meant to compare two sets of kubernetes files stored locally on your computer.
+This is useful when comparing two different methods of deploying an application to kubernetes.
+For example in the case of Grafana Labs, we publish both Helm and Jsonnet for deploying our backend database, Grafana Mimir.
+
+## Motivation
+
+Directly comparing `tanka export` output to `helm template` output imposes several challenges:
+
+1. Resources may be named differently, but otherwise identical.
+   - e.g. A StatefulSet called `mimir-ingester` (Helm) vs `ingester` (Jsonnet)
+2. Resource defaults may be explicitly defined by one set of manifests despite having no impact on the final state
+   - e.g. setting `initContainers: []` in a PodTemplaceSpec.
+3. Some differences may not be of interest to you
+   - e.g. the `helm.sh/chart` label common in Helm charts doesn't really make sense in Jsonnet
+
+Lastly, there may be so many differences that it's not reasonable to fix them all at once. 
+Instead, this tool allows you to fix differences _incrementally_.
+
+The same motivation above applies to comparing Mimir's configuration. There are
+many fields, not all of them are relevant to compare, and the defaults applied
+by Mimir can sometimes obscure the final diff.
+
+## Directory of Tools
+
+| Tool            | Purpose                                                                                                                                        |
+|-----------------|------------------------------------------------------------------------------------------------------------------------------------------------|
+| yaml-patch      | apply JSONPatch documents to yaml files (useful for patching out uninteresting differences)                                                    |
+| k8s-defaulter   | apply kubernetes defaulting logic to yaml files **This requires a running k8s api server, since it works via the server-side dry-run feature** |
+| config-generate | Generate final mimir configuration from a set of kubernetes Deployments or StatefulSets. **This requires docker**                              |
+| config-defaults | Remove fields from mimir configuration that match Mimir defaults. **This requires docker**                                                     |
+
+
+## yaml-patch
+
+### How it works
+
+yaml-patch copies yaml files from one directory to another and optionally performs some transformation to the data in between.
+
+Internally, yaml-patch has two main phases as indicated in the diagram below.
+
+1. Preprocessing
+    - During this phase, manifests are modified in order to eliminate superfluous differences that are deemed unimportant by the user
+2. Validation
+    - During this phase, debug information collected during preprocessing is used to validate that all rules defined in the config file actually impact the final output.
+
+```mermaid
+flowchart LR
```

Comment:
TIL you can do this. amazing

Reply:
It's new! https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/